src/controller.cpp

```diff
@@ -481,25 +483,17 @@ Controller::UpdateResult Controller::update(
     std::clamp(controller_state_.error_integral_ang, -windup_limit, windup_limit);
 
   // Take derivative of error, first the raw unfiltered data:
-  controller_state_.error_deriv_lat.push(
-    (controller_state_.error_lat.errors().at<0>() - controller_state_.error_lat.errors().at<1>()) /
-    dt.toSec());
-  controller_state_.error_deriv_ang.push(
-    (controller_state_.error_ang.errors().at<0>() - controller_state_.error_ang.errors().at<1>()) /
-    dt.toSec());
+  auto error_deriv_lat = controller_state_.error_deriv_lat.filter(error_lat_filtered, dt.toSec());
+  auto error_deriv_ang = controller_state_.error_deriv_ang.filter(error_ang_filtered, dt.toSec());
 
   // calculate the control effort
-  const auto proportional_lat =
-    config_.Kp_lat * controller_state_.error_lat.filtered_errors().at<0>();
+  const auto proportional_lat = config_.Kp_lat * error_lat_filtered;
   const auto integral_lat = config_.Ki_lat * controller_state_.error_integral_lat;
```

Comment:
shall we move this value to just error_integral_lat to use the same structure as the other components?, see previous and next code lines

Reply:
Good idea. I've created a integral filter with unit tests